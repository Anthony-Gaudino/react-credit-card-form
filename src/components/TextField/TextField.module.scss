@use "../../theme.scss";

.textField {
  @include theme.field;
}
