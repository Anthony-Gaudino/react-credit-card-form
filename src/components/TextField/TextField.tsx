import clsx from "clsx";
import { FormEventHandler } from "react";

import InputLabel from "../InputLabel";
import styles from "./TextField.module.scss";

interface TextFieldProps {
  /**
   * Type of input field.
   */
  type: "email" | "number" | "password" | "search" | "tel" | "text" | "url";
  /**
   * Input field label.
   */
  label: string;
  /**
   * If `true` hides the label. Defaults to false;
   */
  hideLabel?: boolean;
  /**
   * Input field value.
   */
  value?: string;
  /**
   * Input field pattern regex.
   */
  pattern?: string;
  /**
   * Input field max string length.
   */
  maxLength?: number;
  /**
   * Input field max string length.
   */
  minLength?: number;
  /**
   * Type of input mode.
   */
  inputMode?:
    | "none"
    | "text"
    | "tel"
    | "url"
    | "email"
    | "numeric"
    | "decimal"
    | "search";
  /**
   * Input field max string length.
   */
  required?: boolean;
  /**
   * If `true` highlits the input if it's invalid.
   */
  highlightInvalid?: boolean;
  /**
   * onChange function callback.
   */
  onChange?: FormEventHandler;
  /**
   * onFocus function callback.
   */
  onFocus?: FormEventHandler;
}

/**
 * A HTML input with a label.
 */
export const TextField = (props: TextFieldProps): JSX.Element => {
  const {
    label,
    hideLabel = false,
    highlightInvalid = false,
    ...others
  } = props;

  const classes = clsx({
    [styles.textField]: true,
    [styles["textField--highlight-invalid"]]: highlightInvalid,
  });

  return (
    <InputLabel label={label} hidden={hideLabel}>
      <input className={classes} {...others}></input>
    </InputLabel>
  );
};
