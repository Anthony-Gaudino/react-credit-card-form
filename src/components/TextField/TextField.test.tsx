import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import TextField from ".";

describe("Test TextField component", () => {
  test("the text field is correctly rendered", () => {
    const { asFragment } = render(<TextField type="text" label="A label" />);

    const label = screen.getByText("A label");

    expect(label).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("can enter text into field", () => {
    const input = "Hello World";

    render(<TextField type="text" label="A label" />);

    const field = screen.getByRole("textbox") as HTMLInputElement;

    userEvent.type(field, input);

    expect(field.value).toBe(input);
  });
});
