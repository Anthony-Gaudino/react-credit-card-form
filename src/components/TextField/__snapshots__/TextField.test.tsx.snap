// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TextField component the text field is correctly rendered 1`] = `
<DocumentFragment>
  <label
    class="inputLabel"
  >
    A label
    <input
      class="textField"
      type="text"
    />
  </label>
</DocumentFragment>
`;
